### PR TITLE
Fix hasOwnProperty bug again and forever

### PR DESCRIPTION
**Pull request by rightfold**
Seems to have been fixed several times before, but somehow still slipped through.

Uses the same check record updates use now.

### CODE REVIEW

File: src/Data/StrMap.js [hunk 1]
```diff
@@ -6,7 +6,7 @@
 exports._copy = function (m) {
   var r = {};
   for (var k in m) {
-    if (m.hasOwnProperty(k)) {
+    if ({}.hasOwnProperty.call(m, k)) {
```
Comment (garyb) [inline src/Data/StrMap.js]: Can we make a variable for `{}` and reuse it throughout please? `{}` does actually go through the process of constructing a new object every time, so if in heavily called code the overhead can be noticeable.

Comment (rightfold) [inline src/Data/StrMap.js]: `var hasOwnProperty = {}.hasOwnProperty;` would also cache the lookup of the method. 👌🏻

File: src/Data/StrMap.js [hunk 2]
```diff
@@ -3,10 +3,12 @@
 
 // module Data.StrMap
 
+var hasOwnProperty = {}.hasOwnProperty;
```
Comment (michaelficarra) [inline src/Data/StrMap.js]: You can just delete this line. The global object already has this binding for you.

File: src/Data/StrMap.js [hunk 3]
```diff
@@ -3,10 +3,12 @@
...
+
 exports._copy = function (m) {
   var r = {};
   for (var k in m) {
-    if (m.hasOwnProperty(k)) {
+    if (hasOwnProperty(m, k)) {
```
Comment (michaelficarra) [inline src/Data/StrMap.js]: You forgot `.call`.

Comment (rightfold) [inline src/Data/StrMap.js]: `_copy` isn't being used at all. Let's delete it.

**Comment by garyb** [comment]
Looks good to me now if @michaelficarra is also happy?

@paf31 any objections? I'm not really sure why previous attempts at this weren't accepted :smile:

File: test/Test/Data/StrMap.purs [hunk 4]
```diff
@@ -34,7 +34,8 @@ instance showInstruction :: (Show k, Show v) => Show (Instruction k v) where
 instance arbInstruction :: (Arbitrary v) => Arbitrary (Instruction String v) where
   arbitrary = do
     b <- arbitrary
-    k <- arbitrary
+    kIshasOwnProperty <- (&&) <$> arbitrary <*> arbitrary
+    k <- if kIshasOwnProperty then pure "hasOwnProperty" else arbitrary
```
Comment (hdgarrood) [inline test/Test/Data/StrMap.purs]: Might `frequency` be a more idiomatic way of saying that we should occasionally use 'hasOwnProperty' as a key? https://pursuit.purescript.org/packages/purescript-quickcheck/3.1.1/docs/Test.QuickCheck.Gen#v:frequency

Comment (rightfold) [inline test/Test/Data/StrMap.purs]: Ya! Didn't know about that one, will add. 👍

**Comment by paf31** [comment]
Looks good to me. I just wanted to avoid the approach where we mangled keys, since `StrMap` needs to continue to be usable as an FFI structure.

Thanks @rightfold!

**Pull request merged**